### ISSUE TITLE
A separate sosnb format.

### ISSUE THREAD

**Issue by BoPeng** [BoPeng/SOS#149]
sos notebook and sos scripts are executed in different ways. A sos notebook are executed cell by cell, potentially by different kernels. A sos script is executed by sos with scripts executed as separate files. The major differences are that

1. Variables in sos scripts are all global, at least for steps without `input` and `output` that are treated by sos kernel as global section. However, these `global sections` can spread to different cells.

2. Scripts belong to the same type (e.g. `with R` or `use R` are executed continuously by the underlying interactive kernel. That is to say, if in one cell with `#with R` and `a=1`, the variable `a` will not be seen by SoS, but will be seen by other cells.

Because of these, how to convert between ipynb to and from sos can be a major problem. I suppose we need a separate `.sosnb` format (or `fileformat=SOSNB1.0`) so that

1. `.ipynb` <-> '.sosnb` would be lossless. We keep markdown, cell number, and keep `#with R` stuff so that an `.ipynb` file can be generated from `.sosnb` just without the result cells.

2. `.ipynb` or `.sosnb` to `.sos` would create valid `.sos` file. The conversion process can try to merge global sections, replace `#with R` with `R:` etc, but there is no guarantee that the generated .sos file is executable because of the global variable issue. 

3. `.sos` to `.ipynb` or `.sosnb` is relatively easier   but we need to decide how to handle section headers.

**Comment by BoPeng** [comment]
I think it can be good enough to

1. Users can use ipython notebook in the current manner, with the `#use R` and `#with R` magic.
2. Users can save such scripts to `.sos`, but SoS cannot execute them because of the blocks of R code etc. SoS can give warnings during conversion.
3. If users would like to convert their conveniently created notebook to a workflow, they would have to manually change 

    ```
    #with R
    R-script
    ```
    to
     ```
    R:
    R-script
    ```
    and make sure the script can be executed as a standalone script (unrelated to other cells). Then the resulting `.sos` file has a bigger chance of being executed as a workflow.
4. `.sos` -> `.ipynb` would mostly be ok, but because the cells are executed one by one, there is no chance for make-style calling (because sos does not know other cells).

In summary, `jupyter/SoS` allows the notebook style use of SoS and can be used to create reports etc. It is then user's responsibility to convert the scratch to a real workflow..

I hope this makes sense.
